--- example7.py	2013-08-09 00:27:31.226209552 -0400
+++ example8.py	2013-08-09 00:27:38.994325870 -0400
@@ -2,18 +2,17 @@
 
 import bs4
 import multiprocessing
-import queue
 import sys
-import threading
 import urllib.request
 
 
 def pipeline(target, i):
-    o = queue.Queue()
+    o = multiprocessing.JoinableQueue()
 
     for _ in range(multiprocessing.cpu_count()):
-        thread = threading.Thread(target=target, args=(i, o), daemon=True)
-        thread.start()
+        proc = multiprocessing.Process(target=target, args=(i, o),
+                                       daemon=True)
+        proc.start()
 
     return o
 
@@ -37,14 +36,16 @@ def extract(i, o):
 
 
 def extract_images(urls):
-    i = queue.Queue()
+    i = multiprocessing.JoinableQueue()
     q = pipeline(urlget, i)
     o = pipeline(extract, q)
 
     for url in urls:
         i.put(url)
 
-    while i.unfinished_tasks > 0 or q.unfinished_tasks > 0 or not o.empty():
+    while (not i._unfinished_tasks._semlock._is_zero() or
+           not i._unfinished_tasks._semlock._is_zero() or
+           not o.empty()):
         yield o.get()
 
 
